File: csharp/Oojs.Tests/TestHelpers.cs
using Oojs;

namespace Oojs.Tests;

internal static class TestHelpers
{
    public static Dictionary<string, object?> MinimalSchema() => new()
    {
        ["$oojs"] = "1.0",
        ["$id"] = "https://example.org/schemas/test",
        ["types"] = new Dictionary<string, object?>
        {
            ["Animal"] = new Dictionary<string, object?>
            {
                ["abstract"] = true,
                ["properties"] = new Dictionary<string, object?>
                {
                    ["name"] = new Dictionary<string, object?> { ["type"] = "string" },
                    ["age"] = new Dictionary<string, object?> { ["type"] = "integer", ["minimum"] = 0 },
                },
                ["required"] = new List<object?> { "name" },
            },
            ["Dog"] = new Dictionary<string, object?>
            {
                ["extends"] = "Animal",
                ["properties"] = new Dictionary<string, object?>
                {
                    ["breed"] = new Dictionary<string, object?> { ["type"] = "string" },
                },
                ["required"] = new List<object?> { "breed" },
            },
            ["Cat"] = new Dictionary<string, object?>
            {
                ["extends"] = "Animal",
                ["properties"] = new Dictionary<string, object?>
                {
                    ["indoor"] = new Dictionary<string, object?> { ["type"] = "boolean" },
                },
            },
        },
    };

    public static (Registry registry, Schema schema) AnimalRegistry()
    {
        var r = new Registry();
        var schema = r.LoadDict(MinimalSchema());
        return (r, schema);
    }

    public static bool HasCode(IEnumerable<ValidationError> errors, string code) =>
        errors.Any(e => e.Code == code);
}

File: php/tests/TestHelpers.php
<?php

declare(strict_types=1);

namespace Oojs\Tests;

use Oojs\Model\Schema;
use Oojs\Registry;
use Oojs\ValidationError;

/**
 * Shared constants and helpers used across all test classes.
 */
trait TestHelpers
{
    protected static function minimalSchema(): array
    {
        return [
            '$oojs' => '1.0',
            '$id'   => 'https://example.org/schemas/test',
            'types' => [
                'Animal' => [
                    'abstract'   => true,
                    'properties' => [
                        'name' => ['type' => 'string'],
                        'age'  => ['type' => 'integer', 'minimum' => 0],
                    ],
                    'required' => ['name'],
                ],
                'Dog' => [
                    'extends'    => 'Animal',
                    'properties' => ['breed' => ['type' => 'string']],
                    'required'   => ['breed'],
                ],
                'Cat' => [
                    'extends'    => 'Animal',
                    'properties' => ['indoor' => ['type' => 'boolean']],
                ],
            ],
        ];
    }

    protected static function makeRegistry(array ...$schemas): Registry
    {
        $r = new Registry();
        foreach ($schemas as $s) {
            $r->loadDict($s);
        }
        return $r;
    }

    /**
     * @return array{Registry, Schema}
     */
    protected static function animalRegistry(): array
    {
        $r      = self::makeRegistry(self::minimalSchema());
        $schema = $r->getSchema('https://example.org/schemas/test');
        return [$r, $schema];
    }

    /**
     * @param ValidationError[] $errs
     */
    protected function hasCode(array $errs, string $code): bool
    {
        foreach ($errs as $e) {
            if ($e->code === $code) {
                return true;
            }
        }
        return false;
    }

    /**
     * @param ValidationError[] $errs
     */
    protected function hasCodeAndMessage(array $errs, string $code, string $fragment): bool
    {
        foreach ($errs as $e) {
            if ($e->code === $code && str_contains($e->message, $fragment)) {
                return true;
            }
        }
        return false;
    }
}
